base/clusterserialize.jl

```diff
@@ -40,9 +50,108 @@ function serialize(s::ClusterSerializer, t::TypeName)
     serialize(s, send_whole)
     write(s.io, identifier)
     if send_whole
+        # Track globals referenced in this anonymous function.
+        # This information is used to resend modified globals when we
+        # only send the identifier.
+        prev = s.anonfunc_id
+        s.anonfunc_id = identifier
         serialize_typename(s, t)
+        s.anonfunc_id = prev
         push!(s.sent_objects, identifier)
+        finalizer(t, x->cleanup_tname_glbs(s, identifier))
     end
-#   println(t.module, ":", t.name, ", id:", identifier, send_whole ? " sent" : " NOT sent")
+
+    # Send global refs if required.
+    syms = syms_2b_sent(s, identifier)
+    serialize(s, syms)
+    foreach(sym->serialize_global_from_main(s, sym), syms)
     nothing
 end
+
+function serialize(s::ClusterSerializer, g::GlobalRef)
+    # Record if required and then invoke the default GlobalRef serializer.
+    sym = g.name
+    if g.mod === Main && isdefined(g.mod, sym)
+        v = getfield(Main, sym)
+        if  !isa(v, DataType) && !isa(v, Module) &&
+            (sym in names(Main, false, false)) && (s.anonfunc_id != 0)
+            # FIXME : There must be a better way to detect if a binding has been imported
+            # into Main or has been primarily defined here.
```

Comment:
Yes, you can use `binding_module(Main, sym) === Main`. Would be good to avoid allocating the `names` list every time.